0006-bingings.md

```diff
@@ -0,0 +1,554 @@
+* If returning the preempted value into the registers exceed the memory container it panics with `MemoryAccessViolation`;
+
+###### Current bugs
+* This function currently does not exist.
+
+---
+```rust
+storage_remove(key_len: u64, key_ptr: u64, register_id: u64)
+```
+Removes the value stored under the given key.
+###### Normal operation
+Very similar to `storage_read`:
+* If key is used, removes the key-value from the trie and copies the content of the value into the `register_id`, even if the content is zero bytes.
+  The respective register is then considered to be used, i.e. `register_len(register_id)` will not return `u64::MAX`.
+* If key is not present then `register_id` is emptied, i.e. `register_len(register_id)` returns `u64::MAX` after this
+  operation.
+
+###### Panics
+* If `key_len + key_ptr` exceeds the memory container or points to an unused register it panics with `MemoryAccessViolation`;
+* If the registers exceed the memory limit panics with `MemoryAccessViolation`;
+* If returning the preempted value into the registers exceed the memory container it panics with `MemoryAccessViolation`;
+
+
+###### Current bugs
+* Does not return into the registers.
+
+---
+```rust
+storage_has_key(key_len: u64, key_ptr: u64) -> u64
+``` 
+Checks if there is a key-value pair.
+###### Normal operation
+* If key is used returns `1`, even if the value is zero bytes;
+* Otherwise returns `0`.
+
+###### Panics
+* If `key_len + key_ptr` exceeds the memory container it panics with `MemoryAccessViolation`;
+
+---
+```rust
+storage_iter_prefix(prefix_len: u64, prefix_ptr: u64) -> u64
+```
+Creates an iterator object inside the host.
+Returns the identifier that uniquely differentiates the given iterator from other iterators that can be simultaneously
+created.
+###### Normal operation
+* It iterates over the keys that have the provided prefix. The order of iteration is defined by the lexicographic
+order of the bytes in the keys. If there are no keys, it creates an empty iterator, see below on empty iterators;
+###### Panics
+* If `prefix_len + prefix_ptr` exceeds the memory container it panics with `MemoryAccessViolation`;
+
+---
+```rust
+storage_iter_range(start_len: u64, start_ptr: u64, end_len: u64, end_ptr: u64) -> u64
+```
+Similarly to `storage_iter_prefix`
+creates an iterator object inside the host.
+###### Normal operation
+Unless lexicographically `start < end`, it creates empty an iterator.
+Iterates over all key-values such that keys are between `start` and `end`, where `start` is inclusive and `end` is exclusive.
+
+Note, this definition allows to either `start` or `end` keys to not actually exist on the given trie.
+
+###### Panics:
+* If `start_len + start_ptr` or `end_len + end_ptr` exceeds the memory container or points to an unused register it panics with `MemoryAccessViolation`;
+
+---
+```rust
+storage_iter_next(iterator_id: u64, key_register_id: u64, value_register_id: u64) -> u64
```

Comment:
why does this return the length of the value but storage_read for example doesn't?

Reply:
`storage_iter_next` does not have to return the length but should be able to signal the following situations upon calling `storage_iter_next`:
* Iterator has advanced to key/value;
* Iterator has stopped.

We can use `u64`-encoded bool to differentiate between them (1 for not stopped iterator, 0 for stopped), but then we also optionally want to know about the following case:
* Iterator has advanced to key/value, but value is zero bytes, so in some cases there is no need to read `value` from the register.

This means we need three values (0 for stopped iterator, 1 for not stopped but non-zero value size, 2 for not stopped but zero value size).

We could have this encoding, but we also could just return the length of the value with `u64::MAX` signifying the stopped iterator.

Reply:
detecting zero byte value doesn't seem like a useful case, and anyway the result of this function is stored in `register_len(value_register_id)` so returning is just for convenience.
but then the question is, why not have a return value in storage_read?
and also, why not `register_len(key_register_id)`

Reply:
Because it is more common when someone write `map` of `filter` over iterator to skip values that are empty. That's why it is the only place where it is useful. But I agree that `register_len` can be used anyway and looks more consistent. So I will change it.

Reply:
Changed.